activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb

```diff
@@ -113,6 +113,25 @@ def db_config
     # * private methods that require being called in a +synchronize+ blocks
     #   are now explicitly documented
     class ConnectionPool
+      class Lease # :nodoc:
```

Comment:
I like the structure of having a new purpose-specific class here, but I do worry that we'll now be creating a new instance of this every time we enter a `with_connection` block in our hoped-for happy path :confused:

Reply:
Yeah, I might as well keep the object and set `lease.connection = nil`

Reply:
Alright, I refactored further and `Lease` are now permanent.